src/Compiler/Service/SynExpr.fs

```diff
@@ -824,6 +858,25 @@ module SynExpr =
                 ->
                 true
 
+            // A match-like construct could be problematically nested like this:
+            //
+            //     match () with
+            //     | _ when
+            //         p &&
+            //         let x = f ()
+            //         (let y = z
+            //          match x with
+            //          | 3 | _ -> y) -> ()
+            //     | _ -> ()
+            | _, Dangling.Match matchOrTry when
+                let line = getSourceLineStr matchOrTry.Range.EndLine
+                let endCol = matchOrTry.Range.EndColumn
+
+                line.Length > endCol + 1
+                && line.AsSpan(endCol + 1).TrimStart(' ').StartsWith("->".AsSpan())
+                ->
+                true
```

Comment:
This technically traverses constructs which will be traversed again later by the default syntax walker. The existing "dangling" calls or matches do the same thing, but they are only triggered in more limited scenarios, whereas this one is run for every (parenthesized) node.

In theory, we could avoid this by doing something like maintaining mutable state throughout the AST traversal and retroactively marking parens that originally looked unnecessary as necessary when we observe that later on, or doing the same with continuation functions, but whether that would actually be faster would depend on the source code, and it would definitely make this code and this function's callsite more complicated.

Alternatively, we could always look at the source line string to check for a trailing arrow first, and only if we find one traverse deeper to look for a dangling match. Since the current implementation of `getSourceLineStr` allocates a new string on every call, though, my guess is that that might actually be significantly slower than just (re)traversing a few extra nodes.

As it is, this only increases time complexity by a constant amount, viz., the number of nodes between a given parenthesized node and the next "break" (paren, brace, or other atomic or delimited construct; anything not covered [here](https://github.com/dotnet/fsharp/blob/39e5a22453bdfb4f62604eeae9a5e03e3fda2edc/src/Compiler/Service/SynExpr.fs#L401-L429)), so I think it's probably fine to leave it. It doesn't seem to make a noticeable difference in speed when working in the FCS project.

Reply:
I would suggest to keep it as you have it.
In practice, allocations are a lot more painful in VS compared to CPU time spent on codefix-related analysis.